### PR TITLE
Using bitwise operation to determine even

### PR DESCRIPTION
**Pull request by Jun10ng**
This PR uses bitwise operation `n&1 == 0` to determine even instead of `n%2 == 0`,  this change brings a very small performance improvement(1%~3%).

PLMK if there is a better way to optimize.

### CODE REVIEW

**Comment by menghanl** [comment]
> performance improvement(1%~3%).

Can you add a benchmark to https://github.com/grpc/grpc-go/blob/master/benchmark/primitives/primitives_test.go and paste the results here? Thanks!

**Comment by Jun10ng** [comment]
run `go test -v -run=primitives_test.go -bench=BenchmarkEven -count=3`

```
BenchmarkEvenBit
BenchmarkEvenBit-8             1        5398731763 ns/op
BenchmarkEvenBit-8             1        5362867605 ns/op
BenchmarkEvenBit-8             1        5394518114 ns/op
BenchmarkEvenMod
BenchmarkEvenMod-8             1        5426930557 ns/op
BenchmarkEvenMod-8             1        5402114883 ns/op
BenchmarkEvenMod-8             1        5425064237 ns/op
PASS
```

**Comment by menghanl** [comment]
Thanks for the benchmark.

@dfawley This change SGTM. Do you have any concerns?

File: benchmark/primitives/primitives_test.go [hunk 1]
```diff
@@ -425,3 +425,26 @@ func BenchmarkRLockUnlock(b *testing.B) {
 		}
 	})
 }
+
+func BenchmarkEvenBit(b *testing.B) {
+	n := int64(1 << 33)
+	b.ResetTimer()
+	for i := int64(0); i <= n; i++ {
+		runBit(i)
+	}
```
Comment (dfawley) [inline benchmark/primitives/primitives_test.go:434]: This isn't how benchmarks are supposed to work in my experience.

1. you are ignoring b.N, so the measurement won't ever converge.
2. you are performing 2^33 operations for each iteration.  Each iteration should only do the operation once, and let b.N determine how many times you do it.
3. `b.ResetTimer()` and `b.StopTimer()` aren't necessary in these situations.

**Comment by dfawley** [comment]
Any gains here are purely theoretical.  We are talking about a 0.5% speedup in the benchmarks (which may be inaccurate given what I called out, above), for something that should only be taking a few nanoseconds either way, and only one is on any kind of critical path (creating headers for an RPC).

In general, I don't think we want to be in the habit of accepting micro optimizations (e.g. ones that wouldn't reflect in RPC-level benchmarks).

**Comment by Jun10ng** [comment]
Thanks for your reply. I will close this issue.

**Pull request closed without merging**